control/controlclient/auto.go

```diff
@@ -46,6 +47,79 @@ func (g *LoginGoal) sendLogoutError(err error) {
 
 var _ Client = (*Auto)(nil)
 
+// waitUnpause waits until the client is unpaused then returns. It only
+// returns an error if the client is closed.
+func (c *Auto) waitUnpause(routineLogName string) error {
+	c.mu.Lock()
+	if !c.paused {
+		c.mu.Unlock()
+		return nil
+	}
+	unpaused := c.unpausedChanLocked()
+	c.mu.Unlock()
+	c.logf("%s: awaiting unpause", routineLogName)
+	select {
+	case <-unpaused:
+		c.logf("%s: unpaused", routineLogName)
+		return nil
+	case <-c.quit:
+		return errors.New("quit")
+	}
+}
+
+// informRoutine is responsible for informing the server of worthy changes to
+// our local state. It runs in its own goroutine.
+func (c *Auto) informRoutine() {
+	defer close(c.informDone)
+	bo := backoff.NewBackoff("informRoutine", c.logf, 30*time.Second)
+	for {
+		if err := c.waitUnpause("informRoutine"); err != nil {
+			c.logf("informRoutine: exiting")
+			return
+		}
+		c.mu.Lock()
+		at := c.lastUpdateAt
+		ctx := c.mapCtx
+		needUpdate := at != c.lastUpdateInformed && !at.IsZero()
+		c.mu.Unlock()
+
+		if needUpdate {
+			select {
+			case <-c.quit:
+				c.logf("informRoutine: exiting")
+				return
+			default:
+			}
+		} else {
+			// Nothing to do, wait for a signal.
+			select {
+			case <-c.quit:
+				c.logf("informRoutine: exiting")
+				return
+			case <-c.informCh:
+				continue
+			}
+		}
+
+		t0 := c.clock.Now()
+		err := c.direct.SendLiteMapUpdate(ctx)
+		d := time.Since(t0).Round(time.Millisecond)
+		if err != nil {
+			if ctx.Err() == nil {
+				c.direct.logf("lite map update after %v: %v", d, err)
+			}
+			bo.BackOff(ctx, err)
```

Comment:
so I worry this can wait up to 30 seconds, even if something else network connectivity-wise has improved/changed

what wakes this up if it turns out the network is back?

Reply:
This is sadly true for the mapRoutine too. There isn't a way to get out of the backoff without canceling the ctx. I can take a stab at that in a follow up change